### ISSUE TITLE
Running instrumented c:geo Tests in Android Studio / strange test & cgeo project structure

### ISSUE THREAD

**Issue by eddiemuc** [cgeo/cgeo#13761]
As of today, I realized that I am no longer able to run instrumented Tests from Android Studio:
* the instrumented test folder "src-android" is not recognized by AS as a test folder
* if I mark it accordingly by myself, it does not found necessary base classes to compile the tests

Running tests via gradle seems to work, but a pure command line output is really not useful in complex debug/development scenarios. Also I don't know how to run only single tests via command line.

While looking into this, two other things became clear:
* Our instrumented tests seems to be written using very very very old JUnit versions! For example they don't even use @Test annotation but still the convention of starting test method names with prefix "test"
* The whole folder structure of the project does not seem to fit to current Android app standards 

I can only speculate that the test execution problems have something to do with the strange and outdated cgeo project.

Folder structure should be something like this (Updated 19.12.22):
```
<root>
  app
    src
      main 
        java -> app sources (java)
        res --> ressources (for main)
      test   -> unit test sources
      androidTest  -> instrumented unit test sources
  other-module
    src
      main  --> module sources
      test
...
```

Folder structure currently is:
```
<root>
  main
    src --> app sources (only java)
    res --> ressources
  other-module
    src --> module sources
  tests
    src --> unit test sources
    src-android --> instrumented unit test sources
```

Questions
* Why do we have this strange structure?
* Would it be beneficial to "clean this up"?

**Comment by ztNFny** [comment]
I opened https://github.com/cgeo/cgeo/issues/13494 for the same or at least a very similar topic recently. Got told in one of our calls that it's due to the way the CI pipeline is set up.
The way it's "supposed to be" is that src-android is only run by CI, not by gradle on a normal build as the tests in there connect against the services, using a user hardcoded in CI pipeline and potentially failing with other users.

So .. "legacy stuff"

**Comment by eddiemuc** [comment]
Note for self (in case a "big move" is necessary in git):
https://vjeko.com/2020/11/24/understanding-renaming-moving-files-with-git/

**Comment by eddiemuc** [comment]
> I opened https://github.com/cgeo/cgeo/issues/13494 for the same or at least a very similar topic recently. Got told in one of our calls that it's due to the way the CI pipeline is set up.
> The way it's "supposed to be" is that src-android is only run by CI, not by gradle on a normal build as the tests in there connect against the services, using a user hardcoded in CI pipeline and potentially failing with other users.
> 
> So .. "legacy stuff"

Yes, have same problem as described in #13494. Even more, if I manually mark the android test folder as "test source" folder in AS, then AS can't compile classes inside (it doesn't find many classes, a bit like what @moving-bits describes for his whole installation)

IMHO this messed up "legacy" structure may well also be the root cause for @moving-bits' problems. I will try around a bit with moving files into standard structure and see if it resolves anything.

**Comment by eddiemuc** [comment]
Structure of a freshly created app example project in AS:
![image](https://user-images.githubusercontent.com/6909759/208386570-b75e97ed-bdc6-466e-a072-ab563f1c9d39.png)

**Comment by eddiemuc** [comment]
For self-reference: what probably needs to be done:
* (Probably optional, but if we're at it anyway) rename folder "main" to "app" 
* move "app/src" to "app/src/main/java"
* move "app/res" to "app/src/main/res"
* move "tests/src" to "app/src/test/java"
* move "tests/src-android" to "app/src/androidTest/java"
* move "tests/res" to "app/src/androidTest/res"
* move files in "tests" (AndroidManifest.xml", "lint.xml", "proguard-rules.xml") to "app/androidTest" -> check if still needed
* Check: if script "tests/scripts/update-trackables" still needed --> anyone knows this?
* delete folder "tests"
* Optional: integrate "app/thirdparty" into "app/src/main/java"

**Comment by bekuno** [comment]
The definition for the both test parts are in the `android.sourceSets` section in the `main/build.gradle`:
https://github.com/cgeo/cgeo/blob/3042c666c6718afe84968e714dd181a68a99f3c5/main/build.gradle#L155-L174

in 687595b7a5b518a4ca649332744742a7d041bd03

**Comment by eddiemuc** [comment]
> The definition for the both test parts are in the `android.sourceSets` section in the `main/build.gradle`:
> https://github.com/cgeo/cgeo/blob/3042c666c6718afe84968e714dd181a68a99f3c5/main/build.gradle#L155-L174
> 
> in 687595b7a5b518a4ca649332744742a7d041bd03
> 

I am aware of that, but AS doesn't seem to get it. And that makes the setting wrong for serious development.

Besides, these settings are very far away from convention, and even if this can be grasped by gradle there should be a good reason for digging into such mists. What is this reason?

**Comment by eddiemuc** [comment]
("legacy" is bot a valid reason)

**Comment by bekuno** [comment]
> Besides, these settings are very far away from convention, and even if this can be grasped by gradle there should be a good reason for digging into such mists. What is this reason?

"Good old history." 

This structure was implemented in "the old times" without Android Studio/gradle, but with Eclipse/ADT.

**Comment by eddiemuc** [comment]
> This structure was implemented in "the old times" without Android Studio/gradle, but with Eclipse/ADT.

The structure recommended by Google + gradle is at least about 20 years old. I personally learned and used this structure with the upcoming of maven back in the days. I used it ever since. The only new thing I see is that there is now a folder `androidTest` beside `main` and `test` in the src folder. Which makes sense since androidTest and (normal unit) test are different things.

So if this is history, it is really ancient history. I think even git did not exist in these times.... No reason to keep it imho.

**Comment by rsudev** [comment]
FYI: The moving around of files led to crowdin translations dropping to ~70%. As the strings are only moved, this can be remedied by e.g. re-importing the translated strings, but it might be worthwhile to notify the translators to not put effort in fixing the situation themselves.
Or what was the plan?

**Comment by eddiemuc** [comment]
> FYI: The moving around of files led to crowdin translations dropping to ~70%. As the strings are only moved, this can be remedied by e.g. re-importing the translated strings, but it might be worthwhile to notify the translators to not put effort in fixing the situation themselves. Or what was the plan?

The plan was for Crowdin to be a stable tool which could handle a simple file move.... 
Well, apart from that: I reimported the cgeo translations also into strings.xml in release branch, and I also updated the changelog-files there. I hope everything is fine now (until crowdin surprises me again).

**Comment by eddiemuc** [comment]
Oh and merry christmas to everyone

**Comment by eddiemuc** [comment]
Update: I added descriptions on how to mave a file and how to sync crowdin in case translations are lost (for whatever reason) to this wiki page: https://github.com/cgeo/cgeo/wiki/Translation-(L18N)

**Comment by rsudev** [comment]
Historically (and perhaps in some contexts still) are files (at least after my experience with translation tools) not a natural or even relevant concept for organizing resources and resource groups for translation. And most translation tools rather 'play safe' and give you the option to re-apply the existing translations from the TM, not doing that automatically by themselves. Can just be a bit tedious for a lot of strings (and I don't know if there is an option in Crowdin to apply all 100% matches with one click).
Thanks for re-importing the translated resources!
I only had to re-apply the translations for strings-old.xml (in German, just a handful, did not yet look after the other languages).

And merry Christmas to you, too!

**Comment by fm-sys** [comment]
Thanks for working on this topic, and merry christmas to everyone!

**Comment by eddiemuc** [comment]
The nightly build is broken since the big move. Apparently it fails because it tries to delete main/res/values/keys.xml after build - which it can't because keys.xml is now at another place.

Should be easy to fix, however I don't have access to the scripts on Jenkins (I have more permissions thanks to @benuko but still can't change jobs apparently). Who can do this?

**Comment by bekuno** [comment]
> Should be easy to fix, however I don't have access to the scripts on Jenkins (I have more permissions thanks to @benuko but still can't change jobs apparently). Who can do this?

I can do it, but later on this day.

**Comment by bekuno** [comment]
> > Should be easy to fix, however I don't have access to the scripts on Jenkins (I have more permissions thanks to @benuko but still can't change jobs apparently). Who can do this?
> 
> I can do it, but later on this day.

In the job `cgeo-CI_PR-build` the script `cgeo build and test for new commits and pull requests v2` was already fine.

In `WatchdogTests_master` I changed the build script.

**Comment by Lineflyer** [comment]
As noted in https://github.com/cgeo/cgeo-status/issues/28 I just also changed some script line for cleanup for the jobs of
- cgeo-nightly
- cgeo-release
- cgeo-release candidate

**Comment by Lineflyer** [comment]
> Should be easy to fix, however I don't have access to the scripts on Jenkins (I have more permissions thanks to @benuko but still can't change jobs apparently). Who can do this?

@eddiemuc 
I checked the access rights and found a typo in your user name, which I corrected. You should now have admin access after authorizing on CI via GitHub.

**Comment by eddiemuc** [comment]
> @eddiemuc I checked the access rights and found a typo in your user name, which I corrected. You should now have admin access after authorizing on CI via GitHub.

Thanks, access works now!

**Comment by eddiemuc** [comment]
While cleaning up the android instrumented tests, I noticed that we have a whole bunch of old Espresso tests (mainly for waypoints plus a bit trackable logging) which are suppressed now for a couple of years. Since they are very old and written in an outdated Espresso format, I doubt that we will ever be able to mae them running again. So as much as it hurts me, I would suggest to delete these tests for good - they are not useful any more.

(For those who don't know it: Espresso is a test framework to perform GUI-driven integration tests)

Those are the tests:
![image](https://user-images.githubusercontent.com/6909759/209539199-3c152b8d-bef0-480c-88be-ed5a74eda756.png)

Are there objections on deleting those?

**Comment by bekuno** [comment]
> Are there objections on deleting those?

+1 to delete them.

**Comment by eddiemuc** [comment]
> > Are there objections on deleting those?
> 
> +1 to delete them.

I read a bit into the topic of Espresso and I believe now I can save some of the tests. 

As part of that I have to remove the last usages of the depreciated lib androidannotations. EditWqPointActivity is the last usage point of this lib in c:geo, and this lib is not compatible with espresso any more it seems.

**Comment by eddiemuc** [comment]
> > > Are there objections on deleting those?
> > 
> > 
> > +1 to delete them.
> 
> I read a bit into the topic of Espresso and I believe now I can save some of the tests.
> 
> As part of that I have to remove the last usages of the depreciated lib androidannotations. EditWqPointActivity is the last usage point of this lib in c:geo, and this lib is not compatible with espresso any more it seems.

Ok I worked on this for many hours last couple of days, and I overcame many hurdles. There's a last hurdle however which is a bit bigger: Espresso is unable to create the EditWaypointActivity (and assumingly many other activities) on the CI because the emulator on the CI has no permission to access fine GPS data (Manifest.permission.ACCESS_FINE_LOCATION). This is because c:geo activities are implemented in such a way that they will hand during onResume in an endless loop asking for such permission over and over again.

While this is a showstopper for GUI test automatization, I also question whether this is legit with respect to c:geo users. I created a new issue for this to discuss: #13798
